scripts/core/editor3/reducers/editor3.tsx

```diff
@@ -89,18 +93,116 @@ export const onChange = (state, newState, force = false) => {
     }
 
     if (force) {
-        return forceUpdate({
-            ...state,
-            editorState,
-        });
+        return forceUpdate(
+            applyAbbreviations({
+                ...state,
+                editorState,
+            })
+        );
     }
 
-    return {
+    return applyAbbreviations({
         ...state,
         editorState,
-    };
+    });
 };
 
+/**
+ * @ngdoc method
+ * @name setAbbreviations
+ * @param {Object} state
+ * @param {Object} abbreviations
+ * @return {Object} returns new state
+ * @description Set the abbreviations dictionary
+ */
+const setAbbreviations = (state, abbreviations) => ({
+    ...state,
+    abbreviations,
+});
+
+/**
+ * @ngdoc method
+ * @name applyAbbreviations
+ * @param {Object} editorState
+ * @param {Array} abbreviations
+ * @return {Object} returns new state
+ * @description Handle the editor tab key pressed event
+ */
+const applyAbbreviations = (state) => {
+    const {editorState, abbreviations} = state;
+    const selection = editorState.getSelection();
+
+    if (!selection.isCollapsed() || abbreviations == null || Object.keys(abbreviations).length === 0) {
+        return state;
+    }
+
+    const content = editorState.getCurrentContent();
+    const block = content.getBlockForKey(selection.getStartKey());
+    const word = getAbbreviationText(block, selection.getStartOffset());
+    const keys = Object.keys(abbreviations);
+    const pattern = '\\b(' + keys.map((item) => escapeRegExp(item)).join('|') + ')(\\*)';
+    const found = word.text.match(new RegExp(pattern, 'g'));
+
+    if (found) {
+        const abbreviation = found[0].replace('*', '');
+        const newWord = abbreviations[abbreviation];
+
+        return replaceWord(state, {word, newWord}, true);
+    }
+
+    return state;
+};
+
+/**
+ * @ngdoc method
+ * @name getAbbreviationText
+ * @param {Object} block
+ * @param {Integer} offset
+ * @return {String} returns text that can contain abbreviation
+ * @description If no current selection from current position extract text that is
+ * delimited on both sides by space or round brackets
+ */
+const getAbbreviationText = (block, offset) => {
+    const text = block.getText();
+    const length = block.getLength();
+    let start = offset;
+    let end = offset;
+
+    while (start > 0 && text[start - 1] !== ' ' && text[start - 1] !== '(' && text[start - 1] !== ')') {
```

Comment:
Why two loops? Couldn't the logic be simpler here. Wouldn't the following work? (pseudocode)

```
for abbreviation in abbreviations
    for match in text.match(abbreviationRegexr)
          if isDeleteSuggestion(block, matchOffset) === false
              replaceWord(...)
```

Reply:
From current position if no selection try to extend on left and right until there is a space, round brackets or delete suggestion.
On editor2 the match is called only once and regex contains all abbreviations. I suppose that this is faster than calling regex for every abbreviation.

Reply:
We could check `if (blockText.includes('*'))` before running the regexp. Should be fast enough to run on the entire document and ignore the browser selection which would make the code simpler.

Reply:
To run something on entire document is a bad idea.
We can check for '*' on the text that possibly contains an abbreviation before to run regexp.

Reply:
> To run something on entire document is a bad idea.

I meant checking the whole document using `if (blockText.includes('*'))` first. It would be fast this way :)